rand_distr/src/hypergeometric.rs

```diff
@@ -0,0 +1,404 @@
+                ln_of_factorial(n1 as f64 - m) +
+                ln_of_factorial(k as f64 - m) +
+                ln_of_factorial((n2 - k) as f64 + m);
+
+            let numerator = (n - k) as f64 * k as f64 * n1 as f64 * n2 as f64;
+            let denominator = (n - 1) as f64 * n as f64 * n as f64;
+            let d = 1.5 * (numerator / denominator).sqrt() + 0.5;
+
+            let x_l = m - d + 0.5;
+            let x_r = m + d + 0.5;
+
+            let k_l = f64::exp(a -
+                ln_of_factorial(x_l) -
+                ln_of_factorial(n1 as f64 - x_l) -
+                ln_of_factorial(k as f64 - x_l) -
+                ln_of_factorial((n2 - k) as f64 + x_l));
+            let k_r = f64::exp(a -
+                ln_of_factorial(x_r - 1.0) -
+                ln_of_factorial(n1 as f64 - x_r + 1.0) -
+                ln_of_factorial(k as f64 - x_r + 1.0) -
+                ln_of_factorial((n2 - k) as f64 + x_r - 1.0));
+            
+            let numerator = x_l * ((n2 - k) as f64 + x_l);
+            let denominator = (n1 as f64 - x_l + 1.0) * (k as f64 - x_l + 1.0);
+            let lambda_l = -((numerator / denominator).ln());
+
+            let numerator = (n1 as f64 - x_r + 1.0) * (k as f64 - x_r + 1.0);
+            let denominator = x_r * ((n2 - k) as f64 + x_r);
+            let lambda_r = -((numerator / denominator).ln());
+
+            // the paper literally gives `p2 + kL/lambdaL` where it (probably)
+            // should have been `p2 <- p1 + kL/lambdaL`; another print error?!
+            let p1 = 2.0 * d;
+            let p2 = p1 + k_l / lambda_l;
+            let p3 = p2 + k_r / lambda_r;
+
+            SamplingMethod::RejectionAcceptance {
+                m, a, lambda_l, lambda_r, x_l, x_r, p1, p2, p3
+            }
+        };
+
+        Ok(Hypergeometric { n1, n2, k, sign_x, offset_x, sampling_method })
+    }
+}
+
+impl Distribution<u64> for Hypergeometric {
+    #[allow(clippy::many_single_char_names)] // Same names as in the reference.
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u64 {
+        use SamplingMethod::*;
+
+        let Hypergeometric { n1, n2, k, sign_x, offset_x, sampling_method } = *self;
+        let x = match sampling_method {
+            InverseTransform { initial_p: mut p, initial_x: mut x } => {
+                let mut u = rng.gen::<f64>();
+                while u > p && x < k as i64 { // the paper erroneously uses `until n < p`, which doesn't make any sense
+                    u -= p;
+                    p *= ((n1 as i64 - x as i64) * (k as i64 - x as i64)) as f64;
+                    p /= ((x as i64 + 1) * (n2 as i64 - k as i64 + 1 + x as i64)) as f64;
+                    x += 1;
+                }
+                x
+            },
+            RejectionAcceptance { m, a, lambda_l, lambda_r, x_l, x_r, p1, p2, p3 } => {
+                loop {
+                    let (y, v) = loop {
+                        let u = rng.gen::<f64>() * p3; // for selecting the region
```

Comment:
This should probably use `gen_range`, or even `Uniform::sample`, with the distribution defined outside the loop.